sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducerClient.cs

```diff
@@ -509,22 +541,64 @@ public virtual Task<string[]> GetPartitionIdsAsync(CancellationToken cancellatio
         ///
         public virtual async Task CloseAsync(CancellationToken cancellationToken = default)
         {
+            cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
             Closed = true;
 
+            var identifier = GetHashCode().ToString();
+
+            // Attempt to close the active transport producers.  In the event that an exception is encountered,
+            // it should not impact the attempt to close the connection, assuming ownership.
+
+            var transportProducerException = default(Exception);
+
             try
             {
-                await InnerProducer.CloseAsync(cancellationToken).ConfigureAwait(false);
+                await GatewayProducer.CloseAsync(cancellationToken).ConfigureAwait(false);
+
+                var pendingCloses = new List<Task>();
+
+                foreach (var producer in PartitionProducers.Values)
+                {
+                    pendingCloses.Add(producer.CloseAsync(CancellationToken.None));
+                }
+
+                PartitionProducers.Clear();
```

Comment:
Just wondering, would it be an issue if someone calls `SendAsync` and we create a new PartitionProducer by calling `PartitionProducers.GetOrAdd(eventBatch.SendOptions.PartitionId, id => Connection.CreateTransportProducer(id, Options));`  while this loop is being executed. Would that new producer becomes a zombie ?

Reply:
Does it make sense to check if the client is closed before calling `PartitionProducers.GetOrAdd(eventBatch.SendOptions.PartitionId, id => Connection.CreateTransportProducer(id, Options));` in the `public SendAsync` method ?

Reply:
The race condition that you're asking about doesn't exist; the collection has its own concurrency guards to prevent multiple calls to the factory.  It will only allow a single factory call and will block others waiting for the producer until the active factory produces one.

Reply:
> Does it make sense to check if the client is closed before calling `PartitionProducers.GetOrAdd(eventBatch.SendOptions.PartitionId, id => Connection.CreateTransportProducer(id, Options));` in the `public SendAsync` method ?

We check at the beginning of the method, but there does exist a benign race condition where `Close` could be called while the send method is running.   What would happen there is that we would potentially end up with a transport producer that doesn't get caught in our "Close all of these" loop.   However, it will be force-closed when the underlying connection is closed  _(everything is linked together)_ or when the idle timeout for the link gets hit.

Even were we to check `IsClosed` before we create one, we'd still the same race condition.  The only way that we'd truly be able to eliminate it would be to apply synchronization across the class, which would degrade performance.   I feel that the benefit of having full determinism over the link versus the complexity and performance impact isn't warranted.   The scenario is corner-case enough that I don't see it as frequent and the impact is minimal since the link will be closed in a short period at worst.